library/alloc/src/boxed/thin.rs

```diff
@@ -91,6 +95,79 @@ impl<T> ThinBox<T> {
 
 #[unstable(feature = "thin_box", issue = "92791")]
 impl<Dyn: ?Sized> ThinBox<Dyn> {
+    #[cfg(not(no_global_oom_handling))]
+    fn new_unsize_zst<T>(value: T) -> Self
+    where
+        T: Unsize<Dyn>,
+    {
+        assert!(mem::size_of::<T>() == 0);
+
+        // This is a helper struct to allocate the header with proper alignment:
+        // Header **end** must be aligned to the value.
+        // We insert the padding not after the `header` field, but before the `header` field.
+        // Allocation layout is be:
+        // ```
+        // [ padding | metadata | ZST value ]
+        // ```
+        #[repr(C)]
+        struct DynZstAlloc<T, Dyn: ?Sized> {
+            header: MaybeUninit<<Dyn as Pointee>::Metadata>,
+            value: MaybeUninit<T>,
+        }
```

Comment:
Why do you need this at all? I would expect you to reuse the layout computation logic that you are already using at runtime.

Reply:
Rewritten this code without using fields.